### PR TITLE
fixed same padding in causal convolutions

### PR DESCRIPTION
**Pull request by Atze00**
# Description

> * Fixed this issue #10062

## Type of change

- [x] Breaking change (Models trained with 3D causal convolution won't work as expected)

## Tests

> * Tested against standard convolution implementation, with different kernels, strides and input shape. This kind of test requires time to run, so it is not included in the pull request.
> * The tests are the same as the one provided originally, the results have been corrected.

## Checklist

- [x] I have signed the [Contributor License Agreement](https://github.com/tensorflow/models/wiki/Contributor-License-Agreements).
- [x] I have read [guidelines for pull request](https://github.com/tensorflow/models/wiki/Submitting-a-pull-request).
- [x] My code follows the [coding guidelines](https://github.com/tensorflow/models/wiki/Coding-guidelines).
- [x] I have performed a self [code review](https://github.com/tensorflow/models/wiki/Code-review) of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

File: official/vision/beta/modeling/layers/nn_layers.py [hunk 1]
```diff
@@ -720,6 +720,7 @@ def _compute_buffered_causal_padding(self,
     Returns:
       A list of paddings for `tf.pad`.
     """
+    shape_in = inputs.shape[1:-1]
```
Comment (Hyperparticle) [inline official/vision/beta/modeling/layers/nn_layers.py:723]: Let's rename to `input_shape`. Also, `inputs.shape` does not work in graph mode, so we should use `tf.shape(inputs)`:

```suggestion
    input_shape = tf.shape(inputs)[1:-1]
```

File: official/vision/beta/modeling/layers/nn_layers.py [hunk 2]
```diff
@@ -730,7 +731,11 @@ def _compute_buffered_causal_padding(self,
          (self.kernel_size[i] - 1) * (self.dilation_rate[i] - 1))
         for i in range(self.rank)
     ]
-    pad_total = [kernel_size_effective[i] - 1 for i in range(self.rank)]
+    pad_total = [max(kernel_size_effective[i] - (self.strides[i]), 0)
```
Comment (Hyperparticle) [inline official/vision/beta/modeling/layers/nn_layers.py:734]: This should only be applied after the first temporal dimension, otherwise the causal padding won't match with `tf.nn.conv1d(padding='causal')`.

```suggestion
    pad_total = [kernel_size_effective[0] - 1]
    for i in range(1, self.rank):
      overlap = (input_shape[i] - 1) % self.strides[i] + 1
      pad_total.append(max(kernel_size_effective[i] - overlap, 0))
```

File: official/vision/beta/modeling/layers/nn_layers_test.py [hunk 3]
```diff
@@ -311,14 +311,14 @@ def test_depthwise_conv3d_causal(self):
...
             [6.0, 6.0, 6.0]],
            [[6.0, 6.0, 6.0],
-            [9.0, 9.0, 9.0]]],
-          [[[8.0, 8.0, 8.0],
+            [4.0, 4.0, 4.0]]],
+          [[[18.0, 18.0, 18.0],
             [12., 12., 12.]],
            [[12., 12., 12.],
-            [18., 18., 18.]]]]])
+            [8., 8., 8.]]]]])
```
Comment (Hyperparticle) [inline official/vision/beta/modeling/layers/nn_layers_test.py:321]: Can you also add these tests?

```python
  def test_conv3d_causal_padding_2d(self):
    """Test to ensure causal padding works like standard padding."""
    conv3d = nn_layers.Conv3D(
        filters=1,
        kernel_size=(1, 3, 3),
        strides=(1, 2, 2),
        padding='causal',
        use_buffered_input=False,
        kernel_initializer='ones',
        use_bias=False,
    )

    keras_conv3d = tf.keras.layers.Conv3D(
        filters=1,
        kernel_size=(1, 3, 3),
        strides=(1, 2, 2),
        padding='same',
        kernel_initializer='ones',
        use_bias=False,
    )

    inputs = tf.ones([1, 1, 4, 4, 1])

    predicted = conv3d(inputs)
    expected = keras_conv3d(inputs)

    self.assertEqual(predicted.shape, expected.shape)
    self.assertAllClose(predicted, expected)

    self.assertAllClose(predicted,
                        [[[[[9.],
                            [6.]],
                           [[6.],
                            [4.]]]]])

  def test_conv3d_causal_padding_1d(self):
    """Test to ensure causal padding works like standard padding."""
    conv3d = nn_layers.Conv3D(
        filters=1,
        kernel_size=(3, 1, 1),
        strides=(2, 1, 1),
        padding='causal',
        use_buffered_input=False,
        kernel_initializer='ones',
        use_bias=False,
    )

    keras_conv1d = tf.keras.layers.Conv1D(
        filters=1,
        kernel_size=3,
        strides=2,
        padding='causal',
        kernel_initializer='ones',
        use_bias=False,
    )

    inputs = tf.ones([1, 4, 1, 1, 1])

    predicted = conv3d(inputs)
    expected = keras_conv1d(tf.squeeze(inputs, axis=[2, 3]))
    expected = tf.reshape(expected, [1, 2, 1, 1, 1])

    self.assertEqual(predicted.shape, expected.shape)
    self.assertAllClose(predicted, expected)

    self.assertAllClose(predicted,
                        [[[[[1.]]],
                          [[[3.]]]]])
```

**Review by Hyperparticle — CHANGES_REQUESTED** [review]
Thanks for the fix! I left some small changes.

File: official/vision/beta/modeling/layers/nn_layers_test.py (hunk 3 above)
Comment (Hyperparticle) [reply official/vision/beta/modeling/layers/nn_layers_test.py:321]: Right below this line, we should also test `_spatial_output_shape`, as it is used in the training loop. Add these lines:

```python
    output_shape = conv3d._spatial_output_shape([4, 4, 4])
    self.assertAllClose(output_shape, [2, 2, 2])
```

File: official/vision/beta/modeling/layers/nn_layers.py (hunk 1 above)
Comment (Hyperparticle) [reply official/vision/beta/modeling/layers/nn_layers.py:723]: The change to requiring the input shape will affect the `_buffered_spatial_output_shape` method. Can we change it to:

```python
  def _buffered_spatial_output_shape(self, spatial_output_shape: List[int]):
    """Computes the spatial output shape from the input shape."""
    # When buffer padding, use 'valid' padding across time. The output shape
    # across time should be the input shape minus any padding, assuming
    # the stride across time is 1.
    if self._use_buffered_input and spatial_output_shape[0] is not None:
      padding = self._compute_buffered_causal_padding(
          tf.zeros([1] + spatial_output_shape + [1]), use_buffered_input=False)
      spatial_output_shape[0] -= sum(padding[1])
    return spatial_output_shape
```

**Review by Hyperparticle — CHANGES_REQUESTED** [review]
I noticed that these changes will break `_compute_causal_padding`, so I added some additional suggestions.

File: official/vision/beta/modeling/layers/nn_layers.py (hunk 2 above)
Comment (Atze00) [reply official/vision/beta/modeling/layers/nn_layers.py]: couldn't this cause to not attend to some of the input also for the standard implementation?

File: official/vision/beta/modeling/layers/nn_layers.py [hunk 4]
```diff
@@ -720,7 +720,7 @@ def _compute_buffered_causal_padding(self,
     Returns:
       A list of paddings for `tf.pad`.
     """
-    shape_in = inputs.shape[1:-1]
+    input_shape = tf.shape(inputs)[1:-1]
     del inputs
```
Comment (Hyperparticle) [inline official/vision/beta/modeling/layers/nn_layers.py:724]: This line isn't necessary, you can remove it.

**Review by Hyperparticle — CHANGES_REQUESTED** [review]

**Review by Hyperparticle — APPROVED** [review]

**Comment by Hyperparticle** [comment]
We might need to add a flag and disable this behavior by default while we wait for new pretrained models.

**Review by yeqingli — APPROVED** [review]

**Review by jaeyounkim — APPROVED** [review]
